Add method release_pointer to vptr

Rbiessy opened the pull request:
The idea is to allow the user to remove pointers without reusing them.
It is particularly useful for sub-buffers because we don't want to keep their virtual pointer. With that it is possible to use Eigen with sub-buffers.

I am not too sure about the method name. Maybe it would make more sense to swap it with the existing remove_pointer (i.e. release_pointer allow to re-use the pointer but remove_pointer does not)?. I'll make sure Eigen synchronises with the latest changes we make here.

Ruyk commented on include/vptr/virtual_ptr.hpp:
```diff
@@ -398,6 +398,13 @@ class PointerMapper {
+   */
+  void release_pointer(const virtual_pointer_t ptr) {
+     m_pointerMap.erase(this->get_node(ptr));
+  }
```
Is not clear to me the difference between release and remove. In fact, for a release I would expect that the buffer is released from the pointer map and returned to the user for their own purposes (like std::unique_ptr::release).
Wouldn't it be better if  we put this functionality on the existing remove_pointer and simply pass a flag to it to not re-use?

Rbiessy replied:
Hum so I don't need the exact same thing as the unique_pre::release. That's because I assume the user already has the buffer.
Anyway I'm fine with adding a flag to remove_pointer to not re-use it. Are you ok with a templated bool parameter? Or do we want a runtime flag for some reason? I think a runtime flag would make sense if we could know whether a buffer is a sub-buffer at runtime but I don't think it's possible. So I would suggest for the template parameter.

Rbiessy replied:
About the tests that's a fair point. I could work on adding a simple one tomorrow maybe.

Rbiessy commented:
I updated the PR with a template bool parameter for remove_pointer instead of release_pointer.
I added 2 more tests, one for the SYCLFree<false> and one for adding an existing pointer. Unfortunately I couldn't reproduce an interesting case with sub-buffers where reusing them causes a crash but I think it makes sense that we don't want to reuse them.

Ruyk commented on tests/vptr/basic.cc:
```diff
@@ -170,6 +170,78 @@ TEST(pointer_mapper, reuse_ptr) {
+    ASSERT_FALSE(PointerMapper::is_nullptr(end));
+    ASSERT_EQ(pMap.count(), 3u);
+
+    // We free the intermediate one and forbit it to be reuse
```
forbit -> forbid

Rbiessy replied:
Oops. Maybe "reused" also?

Ruyk replied:
indeed

Ruyk commented on tests/vptr/basic.cc:
```diff
@@ -170,6 +170,78 @@ TEST(pointer_mapper, reuse_ptr) {
+    ASSERT_FALSE(PointerMapper::is_nullptr(shouldBeNew));
+    ASSERT_EQ(pMap.count(), 3u);
+    ASSERT_NE(shouldBeNew, not_reused);
+  }
```
Could you finish the test by freeing everything that was allocated an ensuring the pMap.count() is back to 0 ?

Ruyk commented on tests/vptr/basic.cc:
```diff
@@ -170,6 +170,78 @@ TEST(pointer_mapper, reuse_ptr) {
+    ASSERT_FALSE(PointerMapper::is_nullptr(shouldBeTheSame));
+    ASSERT_EQ(pMap.count(), 3u);
+    ASSERT_EQ(shouldBeTheSame, reused);
+  }
```
Could you finish the test by freeing everything that was allocated an ensuring the pMap.count() is back to 0 ?

Rbiessy commented on tests/vptr/basic.cc:
```diff
@@ -167,6 +167,99 @@ TEST(pointer_mapper, reuse_ptr) {
+    SYCLfree(end, pMap);
+    ASSERT_EQ(pMap.count(), 1u);
+    SYCLfree(initial, pMap);
+    ASSERT_EQ(pMap.count(), 0u);
```
My tests are based on reuse_ptr, for symmetry I also free'd everything here.

Ruyk replied:
ok, makes sense

Rbiessy commented:
Can I squash and merge this?

Ruyk commented:
Go ahead

Rbiessy commented:
Thanks! I realised too late the name of the commit isn't great, I hope it's not a big deal...
